File: HashTable/599. Minimum Index Sum of Two Lists.py
class Solution:
    def findRestaurant(self, list1: List[str], list2: List[str]) -> List[str]:
        dict = {} #hash map
        for i,n in enumerate(list1):
            if n not in dict:
                dict[n] = i #store the index of list 1
        smallest = []
        smallVal = 2000
        for i,n in enumerate(list2):
            if n in dict:
                if dict[n]+i < smallVal: #greedily check for the smallest
                    smallVal = dict[n]+i
                    smallest = []
                    smallest.append(n)
                elif dict[n]+i == smallVal:
                    smallest.append(n)
        return smallest

File: HashTable/599. Minimum Index Sum of Two Lists.java
import java.util.ArrayList;

class Solution {
    public String[] findRestaurant(String[] list1, String[] list2) {
        //initialize a hash map
        Map<String, Integer> hashmap = new HashMap<>();

        for (int index = 0; index < list1.length; index++) {
            //fill up without duplicates
            if(!hashmap.containsKey(list1[index])){
                hashmap.put(list1[index], index);
            }
        }
        ArrayList<String> smallest = new ArrayList<String>();
        int smallVal = 2000;
        for (int index = 0; index < list2.length; index++) {

            if(hashmap.containsKey(list2[index])){
                if(hashmap.get(list2[index])+index < smallVal){
                    smallVal = hashmap.get(list2[index])+index;
                    smallest.clear();
                    smallest.add(list2[index]);

                }
                else if(hashmap.get(list2[index])+index == smallVal) {
                    smallest.add(list2[index]);
                }
            }
        }
        String[] arrayResult = new String[smallest.size()];
        for (int i = 0; i < smallest.size(); i++) {
            arrayResult[i] = smallest.get(i);
        }
        return arrayResult; 
    }
}